autofit/graphical/messages/transform.py

```diff
@@ -0,0 +1,339 @@
+    return np.log(x) - np.log1p(-x)
+
+
+def sigmoid_grad(x, scale=1, shift=0):
+    expx = np.exp(-x)
+    return scale * expx / np.square(1 + expx)
+
+
+def logit_grad(x, scale=1, shift=0):
+    x = (x - shift) / scale
+    return (np.reciprocal(x) + np.reciprocal(1 - x)) / scale
+
+
+def logit_hess(x, scale=1, shift=0):
+    x = (x - shift) / scale
+    return np.reciprocal(1-x) - np.reciprocal(x)
+
+
+def logit_grad_hess(x, scale=1, shift=0):
+    x = (x - shift) / scale
+    ix = np.reciprocal(x)
+    ix1 = np.reciprocal(1 - x)
+    ix2 = np.square(ix)
+    ix12 = np.square(ix1)
+    return (np.log(x) - np.log1p(-x), (ix + ix1)/scale, (ix12 - ix2)/scale**2)
+
+
+logistic_transform = FunctionTransform(
+    logit, sigmoid, logit_grad, func_grad_hess=logit_grad_hess)
+
+
+def shifted_logistic(shift=0, scale=1):
+    return FunctionTransform(logit, sigmoid, logit_grad, func_grad_hess=logit_grad_hess, args=(scale, shift))
+
+
+def ndtri_grad(x):
+    return np.reciprocal(_norm_pdf(ndtri(x)))
+
+def ndtri_hess(x):
+    return
+
+def ndtri_grad_hess(x):
+    f = ndtri(x)
+    phi = _norm_pdf(f)
+    grad = np.reciprocal(phi)
+    hess = grad**2 * f
+    return f, grad, hess
+
+
+phi_transform = FunctionTransform(
+    ndtri, ndtr, ndtri_grad, func_grad_hess=ndtri_grad_hess)
+
+
+class MultinomialLogitTransform(AbstractDensityTransform):
+    """
+    makes multinomial logististic transform from the p to x, where,
+
+    x_i = log(p_i / (1 - sum(p)))
+    p_i = exp(x_i) / (sum(exp(x_j) for x_j in x) + 1)
+
+    When p's n-simplex is defined by,
+
+    all(0 <= p_i <= 1 for p_i in p) and sum(p) < 1 
+    """
+    def __init__(self, axis=-1):
+        self.axis = axis
+
+    def _validate(self, p):
+        p = np.asanyarray(p)
+        keepdims = np.ndim(p) == self.ndim + 1
+        if not (keepdims or np.ndim(p) == self.ndim):
+            raise ValueError(
+                f"dimension of input must be {self.ndim} or {self.ndim + 1}")
+
+        return p, keepdims
+
+    def transform(self, p):
+        p = np.asanyarray(p)
+        lnp1 = np.log(1 - np.sum(p, axis=self.axis, keepdims=True))
+        lnp = np.log(p)
+        return lnp - lnp1
+
+    def inv_transform(self, x):
+        expx = np.exp(x)
+        return expx / (expx.sum(axis=self.axis, keepdims=True) + 1)
+
+    def jacobian(self, p):
+        p = np.asanyarray(p)
+        pn1 = 1 - np.sum(p, axis=-1, keepdims=True)
+        ln1p = np.log(pn1)
+        lnp = np.log(p)
+        jac = ShermanMorrison(
+            DiagonalMatrix(1/p),
+            1/np.sqrt(pn1) * np.ones_like(p)
+        )
+
+    def log_det(self, p):
+        p = np.asanyarray(p)
+        p1 = 1 - np.sum(p, axis=self.axis, keepdims=True)
+        # Hack to make sure summation broadcasting works correctly
+        log_d = (
+            - np.log(p).sum(axis=self.axis, keepdims=True) - np.log(p1)
+        ) * np.full_like(p, p1.size/p.size)
+        return log_d
+
+    def log_det_grad(self, p):
```

Comment:
These two methods look nearly identical? Consider calling one from the other to avoid repetition

Reply:
yeah, it's difficult to do that without recalculating a value or unnecessarily evaluating a value